pyembed/src/importer.rs

```diff
@@ -900,7 +904,219 @@ impl OxidizedFinder {
             None
         };
 
-        resources_state.pkgutil_modules_infos(py, prefix, state.optimize_level)
+        resources_state.pkgutil_modules_infos(py, prefix, state.optimize_level, |resource| {
+            _PathEntryFinder::is_visible("", &resource.name)
+        })
+    }
+
+    // Canonicalize the path to the current executable or raise an OSError.
+    //
+    // Even `std::env::current_exe()` can fail to `fs::canonicalize()` if, e.g.,
+    // the file is deleted after the program starts.
+    fn exe_realpath(&self, py: Python) -> PyResult<std::path::PathBuf> {
+        let current_exe = &self.state(py).get_resources_state().current_exe;
+        current_exe.canonicalize().map_err(|err| {
+            let exe = current_exe.display();
+            // Use a Python expression instead of PyErr::new to ensure we get the
+            // right OSError subclass.
+            let strerror = format!("cannot open current executable: '{}'", exe);
+            let raw_os_error = err
+                .raw_os_error()
+                .map_or_else(|| "None".to_string(), |err_code| err_code.to_string());
+            let code = if cfg!(windows) {
+                format!(
+                    "OSError(None, r\"{}\", r\"{}\", {})",
+                    strerror, exe, raw_os_error
+                )
+            } else {
+                format!("OSError({}, r\"{}\", r\"{}\")", raw_os_error, strerror, exe)
+            };
+            py.eval(&code, None, None).map_or_else(
+                |err_creating_exc| err_creating_exc,
+                |exc| PyErr::from_instance(py, exc),
+            )
+        })
+    }
+
+    fn path_hook_impl(&self, py: Python, path: PyObject) -> PyResult<_PathEntryFinder> {
+        // Compute _PathEntryFinder::package
+        let pkg = {
+            let current_exe = self.exe_realpath(py)?;
+            let not_exe_err = || {
+                let msg = format!(
+                    "path {} does not begin in \'{}\' (if you're sure it does, check Python's file-system encoding)", &path, current_exe.display());
+                let mut err = PyErr::new::<ImportError, _>(py, msg);
+                let exc = err.instance(py);
+                if let Err(setattr_err) = exc.setattr(py, "path", path.clone_ref(py)) {
+                    setattr_err
+                } else {
+                    err
+                }
+            };
+            let pbuf = pyobject_to_pathbuf(py, path.clone_ref(py))?;
+            let mut p: &std::path::Path = pbuf.as_ref();
+            // path could point "inside" current_exe. If we can't canonicalize,
+            // strip the last component and check again. (zipimporter does
+            // somethings similar)
+            let mut abs_p = p.canonicalize();
```

Comment:
This should use `dunce::canonicalize()` to avoid UNC paths on Windows.

Reply:
Because we check `abs_p == current_exe` later on, what matters is that both are canonicalized using the same function: either both with `dunce` or both with `std`. As discussed above in https://github.com/indygreg/PyOxidizer/pull/343#discussion_r585777519, whether we keep `exe_realpath` or just use `self.state(py).get_resources_state().current_exe` directly, `current_exe` will not necessarily have been canonicalized by `dunce`. So switching this line to using `dunce` would cause `path_hook` to raise `ImportError`s when it should have returned a `_PathEntryFinder`.

Note that _if we keep `exe_realpath`_, then it doesn't matter at all whether we canonicalize using `dunce` or `std`. The canonicalized form is used only to check that `current_exe` matches the head of `abs_p`. `dunce::canonicalize` calls `std::fs::canonicalize` and then does some extra processing that won't be observable. The performance difference surely doesn't matter, but my bias is generally to avoid extra dependencies where they aren't needed.